### DISCUSSION TITLE
Given our small team what should our priorities be?

### DISCUSSION THREAD

**Discussion by jafu888** [androidx/constraintlayout#775] [General]
We have a very small team that support.

- ConstraintLayout Views
- MotionLayout Views
- ConstraintLayout Compose
- MotionLayout Compose

On any of the above we can be working on
- Complete Existing feature gaps (e.g. MotionLayout Compose Parity)
- New Features (ConstraintLayout Views Grid helper) 
- Examples
- Documentation
- Testing

That is a matrix of 20 possible things we could work on. 
Our current priorities:
1. Addressing filed issues/bugs where feasible.
2. Feature complete ConstraintLayout Compose (parity with Views)
3. Feature growth in MotionLayout Compose
4. New Features that help CL/ML API ergonomics in Compose (Grid)
 
 We are working to converge on releasing CL-Compose 1.1.0 and CL 2.2.
 
But More generally what do people think we should work on?

**Comment by jshvarts** [comment]
- MotionLayout Compose feature parity
- More Constraint and Motion layout in Compose codelabs and samples

**Reply by jafu888** [reply]
Complete Parity is far away. 
Any features are critical for what you are trying to do?

We are looking to ideas for small but useful examples we should build.
For MotionLayout XML we built [theses examples](https://developer.android.com/develop/ui/views/animations/motionlayout/examples)

Larger examples  (like [ComposeMail](https://github.com/androidx/constraintlayout/tree/main/demoProjects/ComposeMail)) tended to be too much for developers to understand quickly.

So what examples do people want to see?

**Reply by jshvarts** [reply]
Just in general, would be good to have a decent parity so we can use ML in Compose using json5 knowing that it is basically as capable (or close to it) as ML in the View system

**Reply by jafu888** [reply]
Good to know. 
The engine is common between the two (constraintLayout-core) 
The main issue is connecting all the parameters then testing that they all work.
And documenting it.

The main missing parts in JSON are:

* Multi-Transition management - features involved managing multiple Transitions 
  * Multi-state swipe (Android calculator is a good example)
  * Auto complete, disable
* Transition interloper  - manage the overall easing of a transition
* KeyFrames
  * KeyTimeCycles - low usage, high complexity
  * KeyAttributes/KeyCycles
     * pivot
     * transitionPathRotate

 The DSL still has a lot more to go.

**Comment by ColtonIdle** [comment]
My vote is CL compose parity. all aboard the compose train on my team here. choo choo

**Reply by jafu888** [reply]
ConstraintLayout is very close to parity right now. 
We have this [ConstraintLayout XML to JSON & DSL document](https://github.com/androidx/constraintlayout/wiki/XML--to-JSON-and-DSL-for-ConstraintSet-attributes). Most of what is missing are motionLayout features. 
If you know any gaps let us know.

We doing mostly performance and usability work.
Performance is mostly minimizing the recomposition.
Usability is mostly build examples and asking "how could we make this easier to write?"

 If there are any examples you would like to see let us know. 
 
 MotionLayout is where the big parity gap is.

**Reply by ColtonIdle** [reply]
building constraints in a UI and converting to compose code would be cool. i really don't use CL in compose because I always used the UI for CL in view-land.

**Reply by jafu888** [reply]
Glad to know you would be interested.
We are considering building a tooling like that. After the library is stabilized and goes into beta.

**Comment by wiiznokes** [comment]
A desktop version would be nice

**Reply by jafu888** [reply]
It would not be difficult but it adds another platform to test, validate, and document.
Which will slow our velocity to completion of the core feature set.

**Comment by leinardi** [comment]
My vote is for bugfixing. Currently this bug is preventing to write automated test for some screens: https://issuetracker.google.com/issues/260112195

Btw, should I create an issue also on this GitHub repo?

**Reply by jafu888** [reply]
Does not matter where it is filed. We will find it. 
[260112195](https://issuetracker.google.com/issues/260112195) is being looked at. 
Some Compose bugs are delayed behind bigger refactorings.

You wont see the fixes coming to github as quickly 
Our process was recently changed to be more inline with the rest of androidX
We just released Alpha05

**Comment by lawgimenez** [comment]
How about performance improvements? Some of our apps still supports the minimum SDK of 21 or 24.

**Reply by jafu888** [reply]
For XML we wrote a little overview. https://github.com/androidx/constraintlayout/wiki/ConstraintLayout-performance
And we will be describing how you can track how expensive your layout is. [RelatedCode](https://github.com/androidx/constraintlayout/blob/927e59992067293612bf336dac12956345ff42e7/projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/CheckPerformanceMetric.java#L33)

For Compose most of our current work is about getting it to perform .

If you have specific problems or use cases let us know what is important to you.
